### PR TITLE
remove scanner, rewrite grammar.js

### LINKED ISSUE #7: Odd highlighting for `local-additions`
Some odd highlighting when looking at `help.txt`'s `local-additions` section

w/o TreeSitter |  w/ TreeSitter
:-----:|:-------:
image | image

text:

```help
LOCAL ADDITIONS:				*local-additions*
|nvim-treesitter|  Treesitter configurations and abstraction layer for Neovim.
|packer.txt|                      A use-package inspired Neovim plugin manager
|matchit.txt|	Extended "%" matching

```

notice the `h` in `matchit.txt` is concealed

### LINKED ISSUE #9: parser gets confused by single bar (|)
Example from autocmd.txt:. The `'|'` seems to confuse the parser, and changing it to `"|"` doesn't help. Likely `|` should be treated as a literal if any whitespace follows it before the closing `|`.

```
The special pattern <buffer> or <buffer=N> defines a buffer-local autocommand.
See |autocmd-buflocal|.

Note: The ":autocmd" command can only be followed by another command when the
'|' appears where the pattern is expected.  This works: >
	:augroup mine | au! BufRead | augroup END
But this sees "augroup" as part of the defined command: >
	:augroup mine | au! BufRead * | augroup END
	:augroup mine | au BufRead * set tw=70 | augroup END
Instead you can put the group name into the command: >
	:au! mine BufRead *
	:au mine BufRead * set tw=70
Or use `:execute`: >
	:augroup mine | exe "au! BufRead *" | augroup END
	:augroup mine | exe "au BufRead * set tw=70" | augroup END
```

### LINKED ISSUE #10: tilde "~" headings are not recognized
The `Attributes: ~` subheading in the sample text below is parsed as a normal "line"

```
line [9, 4] - [10, 0]
  word [9, 4] - [9, 15]
  word [9, 16] - [9, 17]
```

sample text:

```
nvim_ui_try_resize({width}, {height})                   *nvim_ui_try_resize()*
    TODO: Documentation

    Attributes: ~
        |RPC| only
```

Maybe there should be `h1`, `h2`, `h3` headings?

- h1:  `========` headings
- h2: `----------` headings
- h3: `Foo ~` headings

### LINKED ISSUE #11: URL token
Seems like the parser currently doesn't recognize URLs in text. The old [makehtml.awk](https://github.com/neovim/neovim/blob/52e4b779e3df1335efc4774e8dfc199d07e24bfd/runtime/doc/makehtml.awk#L325) script had primitive support for that, and it's useful for rendering URLs as hyperlinks.

Example text: https://github.com/neovim/neovim/blob/52e4b779e3df1335efc4774e8dfc199d07e24bfd/runtime/doc/api.txt#L20-L31

```
==============================================================================
API Usage                                               *api-rpc* *RPC* *rpc*

                                                        *msgpack-rpc*
RPC is the typical way to control Nvim programmatically.  Nvim implements the
MessagePack-RPC protocol:
  https://github.com/msgpack-rpc/msgpack-rpc/blob/master/spec.md
  https://github.com/msgpack/msgpack/blob/0b8f5ac/spec.md

Many clients use the API: user interfaces (GUIs), remote plugins, scripts like
"nvr" (https://github.com/mhinz/neovim-remote).  Even Nvim itself can control
other Nvim instances.  API clients can:

  - Call any API function
  - Listen for events
  - Receive remote calls from Nvim
```

### LINKED ISSUE #12: list items (and nested list items)
In order for `:help` to render as something other than fixed-width, hard-wrapped text, (e.g. for [HTML format](https://neovim.io/doc/user/api.html)) the parser needs to give some basic hints about structure besides "line". Sometimes (often) the `:help` source needs to be fixed (because it can't be reasonably parsed), but in other common cases it's probably reasonable for tree-sitter-vimdoc to provide some basic idea of:

1. paragraph (and "nested" paragraphs? AKA indentation) https://github.com/vigoux/tree-sitter-vimdoc/issues/13
2. list items (which may be nested) [this issue]

## Example of a basic list
 https://github.com/neovim/neovim/blob/52e4b779e3df1335efc4774e8dfc199d07e24bfd/runtime/doc/api.txt#L25-L34

```
Many clients use the API: user interfaces (GUIs), remote plugins, scripts like
"nvr" (https://github.com/mhinz/neovim-remote).  Even Nvim itself can control
other Nvim instances.  API clients can:

  - Call any API function
  - Listen for events
  - Receive remote calls from Nvim

The RPC API is like a more powerful version of Vim's "clientserver" feature.
```

## Example of a nested list

api.txt has many nested lists. Maybe it's not worth supporting these? 

https://github.com/neovim/neovim/blob/52e4b779e3df1335efc4774e8dfc199d07e24bfd/runtime/doc/api.txt#L3257-L3273

```
                 • callback (function|string) optional: if a string, the name
                   of a Vimscript function to call when this autocommand is
                   triggered. Otherwise, a Lua function which is called when
                   this autocommand is triggered. Cannot be used with
                   {command}. Lua callbacks can return true to delete the
                   autocommand; in addition, they accept a single table
                   argument with the following keys:
                   • id: (number) the autocommand id
                   • event: (string) the name of the event that triggered the
                     autocommand |autocmd-events|
                   • group: (number|nil) the autocommand group id, if it
                     exists
                   • match: (string) the expanded value of |<amatch>|
                   • buf: (number) the expanded value of |<abuf>|
                   • file: (string) the expanded value of |<afile>|
                   • data: (any) arbitrary data passed to
                     |nvim_exec_autocmds()|
```

### LINKED ISSUE #13: paragraphs (and nested paragraphs?)
_Preamble from https://github.com/vigoux/tree-sitter-vimdoc/issues/12 :_

> In order for `:help` to render as something other than fixed-width, hard-wrapped text, (e.g. for [HTML format](https://neovim.io/doc/user/api.html)) the parser needs to give some basic hints about structure besides "line". Sometimes (often) the `:help` source needs to be fixed (because it can't be reasonably parsed), but in other common cases it's probably reasonable for tree-sitter-vimdoc to provide some basic idea of:

1. paragraph (and "nested" paragraphs? AKA indentation) [this issue]
2. list items (which may be nested) https://github.com/vigoux/tree-sitter-vimdoc/issues/12

## Example of paragraphs and nested (indented) paragraphs

https://github.com/neovim/neovim/blob/52e4b779e3df1335efc4774e8dfc199d07e24bfd/runtime/doc/eval.txt#L51-L74

By "nested paragraphs" I simply mean blocks of text that have common indentation.  In the following example, 

- we can't parse the `Number 123...` block but if we can at least recognize that it is indented, then the [HTML result](https://neovim.io/doc/user/api.html) can wrap that block in `<pre>` so that it is readable.
- the `*octal*` line doesn't share indentation so it should be parsed as a normal "line", possibly part of the next paragraph.
- it may be reasonable for us to require blank lines between these blocks.

```
The Number and String types are converted automatically, depending on how they
are used.

Conversion from a Number to a String is by making the ASCII representation of
the Number.  Examples:
	Number 123	-->	String "123" ~
	Number 0	-->	String "0" ~
	Number -1	-->	String "-1" ~
							*octal*
Conversion from a String to a Number is done by converting the first digits to
a number.  Hexadecimal "0xf9", Octal "017" or "0o17", and Binary "0b10"
numbers are recognized.  If the String doesn't start with digits, the result
is zero. Examples:
	String "456"	-->	Number 456 ~
	String "6bar"	-->	Number 6 ~
	String "foo"	-->	Number 0 ~
	String "0xf1"	-->	Number 241 ~
	String "0100"	-->	Number 64 ~
	String "0o100"	-->	Number 64 ~
	String "0b101"	-->	Number 5 ~
	String "-8"	-->	Number -8 ~
	String "+8"	-->	Number 0 ~

To force conversion from String to Number, add zero to it: >
	:echo "0100" + 0
<	64 ~

To avoid a leading zero to cause octal conversion, or for using a different
base, use |str2nr()|.
```

### LINKED ISSUE #14: single-char 'x' is not an option hotlink
Noticed this while working on gen_help_html.lua https://github.com/neovim/neovim/pull/11967 .

This is just for reference; as mentioned in https://github.com/vigoux/tree-sitter-vimdoc/issues/11#issuecomment-1242980794 I'm pretty deep into rewriting tree-sitter-vimdoc. Results are promising so far, but will take time.

## Problem

The parser currently thinks that 'x' is a link to an option, but that is not correct. single-char surrounded by single-quotes should be treated as a plain word.

```
  ["'!'"] = true,
  ["'#'"] = true,
  ["'$'"] = true,
  ["'%'"] = true,
  ["'&'"] = true,
  ["'('"] = true,
  ["')'"] = true,
  ["'+'"] = true,
  ["','"] = true,
  ["','sqlKeyword'"] = true,
  ["','sqlOption'"] = true,
  ["','sqlStatement'"] = true,
  ["'-'"] = true,
  ["'-'Z'"] = true,
  ["'.'"] = true,
  ["'/'"] = true,
  ["'0'"] = true,
  ["'1'"] = true,
  ["'2'"] = true,
  ["'4'"] = true,
  ["'8'"] = true,
  ["'9'"] = true,
  ["':'"] = true,
  ["';'"] = true,
  ["'<'"] = true,
  ["'='"] = true,
  ["'>'"] = true,
  ["'?'"] = true,
  ["'@'"] = true,
  ["'A'"] = true,
  ["'B'"] = true,
  ["'C'"] = true,
  ["'D'"] = true,
  ["'E'"] = true,
  ["'F'"] = true,
  ["'G'"] = true,
  ["'H'"] = true,
  ["'I'"] = true,
  ["'J'"] = true,
  ["'K'"] = true,
  ["'L'"] = true,
  ["'M'"] = true,
  ["'N'"] = true,
  ["'O'"] = true,
  ["'P'"] = true,
  ["'Q'"] = true,
  ["'R'"] = true,
  ["'S'"] = true,
  ["'T'"] = true,
  ["'W'"] = true,
  ["'X'"] = true,
  ["'Y'"] = true,
  ["'Z'"] = true,
  ["'['"] = true,
  ["']'"] = true,
  ["'^'"] = true,
  ["'_'"] = true,
  ["'a'"] = true,
  ["'b'"] = true,
  ["'c'"] = true,
  ["'d'"] = true,
  ["'e'"] = true,
  ["'f'"] = true,
  ["'g'"] = true,
  ["'h'"] = true,
  ["'i'"] = true,
  ["'j'"] = true,
  ["'k'"] = true,
  ["'l'"] = true,
  ["'m'"] = true,
  ["'n'"] = true,
  ["'o'"] = true,
  ["'p'"] = true,
  ["'q'"] = true,
  ["'r'"] = true,
  ["'s'"] = true,
  ["'t'"] = true,
  ["'u'"] = true,
  ["'v'"] = true,
  ["'w'"] = true,
  ["'x'"] = true,
  ["'y'"] = true,
  ["'z'"] = true,
  ["'~'"] = true,
  ["'·'"] = true,
  ["'ö'"] = true,
  ["'─'"] = true,
  ["'│'"] = true,
  ["'├'"] = true,
  ["'┤'"] = true,
  ["'┬'"] = true,
  ["'┴'"] = true,
  ["'┼'"] = true,
  ["'보'"] = true,
  ["'세'"] = true,
  ["'여'"] = true,
  ["'요'"] = true,
```

### LINKED ISSUE #17: Inline code highlighting has issues
There are multiple issues of recognizing inline code blocks. It usually works fine when backticks surround single word consisting from letters. Other than that, not so much.

Text:
```
`inline code with s`
`inline code with spaces`
`require('')`
`require('mod')`
`{}`
```

Current behavior with tree-sitter:
screenshot_2022-09-26_15:46:20

Current behavior without tree-sitter:
screenshot_2022-09-26_15:46:35

### PR DESCRIPTION
**Pull request by justinmk**
- benefits:
    - introduce `url` and `block` (group of `line`s)
    - fixes most open issues
    - no C scanner
    - faster
    - doesn't hang on `filetype.txt` , `usr_24.txt`
- todo:
    - keycodes https://github.com/vigoux/tree-sitter-vimdoc/issues/1
    - `[range]` https://github.com/vigoux/tree-sitter-vimdoc/issues/1
    - highlights
        - debug w/ `vim.treesitter.get_captures_at_cursor()`
    - ~~code blocks~~
    - ~~fixup remaining test failures~~

fix #7
fix #9
fix #10
fix #11
fix #14
fix #12 (non-nested)
fix #13 (non-nested)
fix #17

### CODE REVIEW

File: corpus/headline.txt [hunk 1]
```diff
@@ -14,11 +14,12 @@ Text
...
       (word)))
-  (line
-    (word)))
+  (block_end
+    (line_noeol
+      (word))))
 
 ================================================================================
-Headline with tag at the start
+headline with tag at the start
```
Comment (justinmk) [inline corpus/headline.txt:22]: use lowercase descriptions so that `tree-sitter test --filter headline` is easy to use.

**Comment by clason** [comment]
To make the tests run, simply remove the argument from https://github.com/vigoux/tree-sitter-vimdoc/blob/5be26866a7fea022ad7efe9a521e953eeb507c0d/Makefile#L2

(ABI 14 is the default now.)

**Comment by justinmk** [comment]
> Incompatible language version 14. Expected minimum 13, maximum 13

@vigoux any idea how I can fix this? Is CI installing an older version of tree-sitter CLI?

**Comment by justinmk** [comment]
`code_block` is kinda working but still needs work. Other than that, all tests are passing (plus several new tests).

**Comment by vigoux** [comment]
It is pretty surprising... I'll look into this problem today.

**Comment by vigoux** [comment]
I pushed a fix for that in latest master.

To ensure consistency of the parser version, I recommend you to:
- Run `npm install`
- Run `make`

This should handle all the version management for us now.

**Comment by clason** [comment]
Yes, looks like CI installs an old parser (0.20.0)?

As ABI 14 is now (0.20.7) the default, it would be better to fix CI.

**Comment by vigoux** [comment]
The CI uses the locked version of tree-sitter.

In latest master the locked version now supports ABI 14, and furthermore, the default ABI for this parser should now be 14.

**Comment by clason** [comment]
Any interest in including queries in this repo, too, so CI can catch breaking changes that require query updates on PRs?

(Does this PR require query updates?)

**Comment by vigoux** [comment]
Would be good too.
I'll have to look up the names of the captures for markup languages.

**Comment by clason** [comment]
You can start with https://github.com/nvim-treesitter/nvim-treesitter/blob/master/queries/help/highlights.scm

(sans the quote conceal). We don't need to follow the strict corset of nvim-treesitter, but it would help; the allowed text captures are here: https://github.com/nvim-treesitter/nvim-treesitter/blob/master/CONTRIBUTING.md#text

**Comment by justinmk** [comment]
> (Does this PR require query updates?)

yeah, mostly the introduction of (block) (which contains (line)s).

I will probably alias `(line_noeol)` => `(line)` and `(block_end)` => `(block)` since there isn't much reason to expose those.

File: README.md [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+---------
+
+- quasi-spec: https://github.com/nanotee/vimdoc-notes
+
+Known issues
+------------
+
+`block_end` and `line_noeol` are intentionally _not_ alised to `block` and `line` until all they are
+a bit more stable. Currently their behavior is not perfect, so having the non-aliased forms in the
+test files helps notice their quirks.
```
Comment (vigoux) [inline README.md:20]: While this is a good idea to say it, having that for a parser version that might be shipped with a released version of neovim seems like a bad idea.

Why not just aliasing it ?

File: grammar.js [hunk 3]
```diff
@@ -1,58 +1,149 @@
...
+    block_end: ($) => choice(
+      choice($.line_noeol, $.line_li_noeol),
+      seq(repeat1(choice($.line, $.line_li)), choice($.line_noeol, $.line_li_noeol))),
+
+    // Code block: adjacent lines surrounded by ">" and "<".
+    // TODO: expose (word)s in code (line)s? But whitespace must be preserved.
+    code_block: ($) => seq(
+      />[\t ]*\n/,
+      repeat(alias($.line_code, $.line1)),
+      alias($.line_code_noeol, $.line2),
```
Comment (vigoux) [inline grammar.js:45]: Can't those two be renamed to `line` ?

File: README.md (hunk 2 above)
Comment (justinmk) [reply README.md:20]: Will alias them. Things have settled down now :)

File: grammar.js (hunk 3 above)
Comment (justinmk) [reply grammar.js:45]: yes, will alias these to "line" now that `code_block` is working.

**Comment by justinmk** [comment]
`code_block` is working. Only 2 failing tests remaining for "option". Then will add `keycode` and `keychord`, then this is ready I think.

**Comment by clason** [comment]
> If we're open to renaming things, I suggest:

We are. If you add the queries here, I'll just plonk them into `nvim-treesitter` when bumping this parser. (Makes my life much easier.)

**Comment by clason** [comment]
> `code_block` is working.

I know it's a long shot (especially now that there's no custom scanner anymore), but any chance the new `code_block` can specify the language (either Lua or Vimscript would be enough)?

**Comment by justinmk** [comment]
> any chance the new `code_block` can specify the language (either Lua or Vimscript

Isn't that something that can be added later? Could also extend the vimdoc spec as @vigoux mentioned, e.g. `vimscript>` and `lua>`.

**Comment by clason** [comment]
Oh, sure, I just wanted to raise awareness in case it matters.

**Comment by vigoux** [comment]
By the way, I am okay with renaming things. I think taking a fresh start is great.

**Comment by clason** [comment]
> Isn't that something that can be added later? Could also extend the vimdoc spec as @vigoux mentioned, e.g. `vimscript>` and `lua>`.

That is a good point; adding `vim>` and `lua>` as valid (concealed) markers to the legacy syntax file would allow us to use these markers in our own help files at least. And once this parser has parity with the syntax file, enable it by default and presto: highlighted code blocks.

**Comment by vigoux** [comment]
For the record, this seems to fail parsing some options.

**Comment by justinmk** [comment]
> For the record, this seems to fail parsing some options.

Yes, I need to fix some stuff

**Comment by justinmk** [comment]
options are fixed, now 2 other issues to fix :D getting close...

**Comment by justinmk** [comment]
Ok all tests are passing and I think the results are pretty good. I will note the compromises as PR comments.

File: corpus/arguments.txt [hunk 4]
```diff
@@ -1,31 +1,52 @@
...
+{foo "{bar}" `{baz}` |{baz| }
+
+--------------------------------------------------------------------------------
+
+(help_file
+  (block
+    (line
+      (argument
+        (word)
+        (ERROR))
```
Comment (justinmk) [inline corpus/arguments.txt:46]: I gave up on this for now, ideally it should be:

      (argument
        (word)
        (MISSING "}"))

File: corpus/backtick.txt [hunk 5]
```diff
@@ -1,45 +1,83 @@
...
+      (backtick
+        (word))
+      (word)
+      (word)
+      (word)
+      (backtick
+        (word))
+      (word)
+      (word)
+      (ERROR
```
Comment (justinmk) [inline corpus/backtick.txt:41]: Perhaps this should be:

      (backtick
        (word)
        (MISSING "`")))

but arguably, ERROR is correct when someone opens an unclosed (on the same line) backtick?

File: corpus/backtick.txt [hunk 6]
```diff
@@ -1,45 +1,83 @@
...
+      (backtick
+        (word)))))
 
 ================================================================================
-Backtick in text
+backtick in text
 ================================================================================
 
-Hello `world`, I am a markup language
+Hello `world`, I am `markup language`. But `this is
```
Comment (justinmk) [inline corpus/backtick.txt:23]: The grammar supports same-line whitespace in a codespan, but I'm not sure if that's actually supported by vimdoc format.

Comment (justinmk) [reply corpus/backtick.txt:23]: never mind this is fixed by inlining the `$.word2` cases into `$.word`. 🤷

File: corpus/backtick.txt (hunk 5 above)
Comment (justinmk) [reply corpus/backtick.txt]: never mind this is fixed by inlining the `$.word2` cases into `$.word`. 🤷

**Comment by clason** [comment]
> I didn't rename anything yet, nor add highlights. Should that wait for a followup PR?

Best to rip the bandaid off in one go.

And, yes, please add queries; it's good practice to keep them in sync, otherwise I'll have to manually fix the ones we bundle or it will block all updates in nvim-treesitter. (Can be improved later; the main point is that they don't error out with the new parser.)

**Comment by justinmk** [comment]
# Plan

1. fix `column_heading` to _not_ parse special forms like links, tags, etc. It only supports plaintext.
2. rename atoms
3. add queries/highlights
4. in a future PR I will try to eliminate the `block_end` and `line_noeol` junk, just for elegance.

**Comment by justinmk** [comment]
Latest push includes renamed forms. See updated comment https://github.com/vigoux/tree-sitter-vimdoc/pull/16#issuecomment-1250864587 for summary.

**Comment by vigoux** [comment]
You should have commit access now, feel free to merge this PR and take the parser over from here

**Pull request merged**